test/e2e/controller_test.go

```diff
@@ -0,0 +1,401 @@
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+
+	catalogv1alpha1 "github.com/kcp-dev/catalog/api/v1alpha1"
+)
+
+// The tests in this package expect to be called when:
+// - kcp is running
+// - a kind cluster is up and running
+// - it is hosting a syncer, and the SyncTarget is ready to go
+// - the controller-manager from this repo is deployed to kcp
+// - that deployment is synced to the kind cluster
+// - the deployment is rolled out & ready
+//
+// We can then check that the controllers defined here are working as expected.
+
+var workspaceName string
+
+func init() {
+	rand.Seed(time.Now().Unix())
+	flag.StringVar(&workspaceName, "workspace", "", "Workspace in which to run these tests.")
+}
+
+func parentWorkspace(t *testing.T) logicalcluster.Name {
+	flag.Parse()
+	if workspaceName == "" {
+		t.Fatal("--workspace cannot be empty")
+	}
+
+	return logicalcluster.New(workspaceName)
+}
+
+func loadClusterConfig(t *testing.T, clusterName logicalcluster.Name) *rest.Config {
+	t.Helper()
+	restConfig, err := config.GetConfigWithContext("base")
+	if err != nil {
+		t.Fatalf("failed to load *rest.Config: %v", err)
+	}
+	return rest.AddUserAgent(kcpclienthelper.SetCluster(rest.CopyConfig(restConfig), clusterName), t.Name())
+}
+
+func loadClient(t *testing.T, clusterName logicalcluster.Name) client.Client {
+	t.Helper()
+	scheme := runtime.NewScheme()
+	if err := clientgoscheme.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add client go to scheme: %v", err)
+	}
+	if err := tenancyv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", tenancyv1alpha1.SchemeGroupVersion, err)
+	}
+	if err := catalogv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", catalogv1alpha1.GroupVersion, err)
+	}
+	if err := apisv1alpha1.AddToScheme(scheme); err != nil {
+		t.Fatalf("failed to add %q to scheme: %v", apisv1alpha1.SchemeGroupVersion, err)
+	}
+	tenancyClient, err := client.New(loadClusterConfig(t, clusterName), client.Options{Scheme: scheme})
+	if err != nil {
+		t.Fatalf("failed to create a client: %v", err)
+	}
+	return tenancyClient
+}
+
+func createWorkspace(t *testing.T, clusterName logicalcluster.Name) client.Client {
+	t.Helper()
+	parent, ok := clusterName.Parent()
+	if !ok {
+		t.Fatalf("cluster %q has no parent", clusterName)
+	}
+	c := loadClient(t, parent)
+	t.Logf("creating workspace %q", clusterName)
+	if err := c.Create(context.TODO(), &tenancyv1alpha1.ClusterWorkspace{
```

Comment:
Global comment: don't use `context.TODO()` unless there is absolutely no way to get access to a preexisting context. The pattern we've established in the kcp e2e tests is that each `TestXYZ()` function does this near the top:

```go
ctx, cancel := context.WithCancel(context.Background())
t.Cleanup(cancel)
`

If you need both `t` and `ctx` in a helper function, you can pass them both in.